### PR TITLE
FINERACT-1724: Creating accrual entry should not be done by native queries

### PR DESCRIPTION
**Pull request by abraham-menyhart**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualWritePlatformServiceImpl.java [hunk 1]
```diff
@@ -265,14 +265,11 @@ private void addAccrualAccounting(LoanScheduleAccrualData scheduleAccrualData, B
...
-        String transactionSql = "INSERT INTO m_loan_transaction  (loan_id,office_id,is_reversed,external_id,transaction_type_enum,transaction_date,amount,interest_portion_derived,"
-                + "fee_charges_portion_derived,penalty_charges_portion_derived, submitted_on_date, created_by, last_modified_by, created_on_utc, last_modified_on_utc) "
-                + "VALUES (?, ?, false, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
-        this.jdbcTemplate.update(transactionSql, scheduleAccrualData.getLoanId(), scheduleAccrualData.getOfficeId(),
-                externalIdFactory.create().getValue(), LoanTransactionType.ACCRUAL.getValue(), accruedTill, amount, interestPortion,
-                feePortion, penaltyPortion, DateUtils.getBusinessLocalDate(), user.getId(), user.getId(),
-                DateUtils.getOffsetDateTimeOfTenantWithMostPrecision(), DateUtils.getOffsetDateTimeOfTenantWithMostPrecision());
-        final Long transactionId = this.jdbcTemplate.queryForObject("SELECT " + sqlGenerator.lastInsertId(), Long.class); // NOSONAR
+
+        Loan loan = loanRepositoryWrapper.findOneWithNotFoundDetection(scheduleAccrualData.getLoanId());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualWritePlatformServiceImpl.java]: Please do not load the full loan here!
The LoanScheduleAccrualData contains the loanId and the officeId, only load the references of them please!

**Review by galovics — CHANGES_REQUESTED** [review]
I agree with Adam's point, the current way could hurt the performance.

In addition, can you pls trigger a new business(external) event when the transaction got created? I think we have a LoanTransactionCreatedBusinessEvent or something similar.

**Review by galovics — APPROVED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualWritePlatformServiceImpl.java [hunk 2]
```diff
@@ -265,14 +270,11 @@ private void addAccrualAccounting(LoanScheduleAccrualData scheduleAccrualData, B
...
-        this.jdbcTemplate.update(transactionSql, scheduleAccrualData.getLoanId(), scheduleAccrualData.getOfficeId(),
-                externalIdFactory.create().getValue(), LoanTransactionType.ACCRUAL.getValue(), accruedTill, amount, interestPortion,
-                feePortion, penaltyPortion, DateUtils.getBusinessLocalDate(), user.getId(), user.getId(),
-                DateUtils.getOffsetDateTimeOfTenantWithMostPrecision(), DateUtils.getOffsetDateTimeOfTenantWithMostPrecision());
-        final Long transactionId = this.jdbcTemplate.queryForObject("SELECT " + sqlGenerator.lastInsertId(), Long.class); // NOSONAR
+        Loan loan = loanRepository.getReferenceById(scheduleAccrualData.getLoanId());
+        Office office = officeRepository.getReferenceById(scheduleAccrualData.getOfficeId());
+        LoanTransaction loanTransaction = LoanTransaction.accrueTransaction(loan, office, accruedTill, amount, interestPortion, feePortion,
+                penaltyPortion, externalIdFactory.create());
+        final Long transactionId = loanTransactionRepository.saveAndFlush(loanTransaction).getId();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanAccrualWritePlatformServiceImpl.java]: Would you please rework the
```
Optional<LoanTransaction> loanAccrualTransaction = loanTransactionRepository.findByIdAndLoanId(transactionId,
                scheduleAccrualData.getLoanId());
        if (!loanAccrualTransaction.isEmpty()) {
            businessEventNotifierService
                    .notifyPostBusinessEvent(new LoanAccrualTransactionCreatedBusinessEvent(loanAccrualTransaction.get()));
        }
 ```       
  to not fetch the Loan transaction from database, rather use the LoanTransaction entity you have just saved.
 
  It is in line 297-302

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**